paasta_tools/kubernetes_tools.py

```diff
@@ -2049,3 +2099,25 @@ def get_pod_hostname(kube_client: KubeClient, pod: V1Pod) -> str:
         return pod.spec.node_name
     # if label has disappeared (say we changed it), default to node name
     return node.metadata.labels.get("yelp.com/hostname", pod.spec.node_name)
+
+
+def to_node_label(label: str) -> str:
+    """k8s-ifies certain special node labels"""
+    if label in {"instance_type", "instance-type"}:
+        # kube_client is assumed to already be configured
+        version_info = kube_client.VersionApi().get_code()
+        if int(version_info.major) == 1 and int(version_info.minor) < 17:
+            # the beta instance-type label is deprecated as of k8s v1.17
+            return "beta.kubernetes.io/instance-type"
+        else:
+            return "node.kubernetes.io/instance-type"
+    elif label in {
+        "datacenter",
```

Comment:
Are these also all configurable via deploy_whitelist/blacklist?

I also wonder if this function is necessary, especially if we need to make an api call to get the kubernetes version. What if we replaced all the old beta selectors with the new name when we do an upgrade?

Reply:
I wanted to create a single function to serve as a source of truth for any sort of label name changing, so it includes the instance-type label fudging, as well as the location label stuff too. deploy_whitelist/blacklist stuff use this function to fudge the location labels.

deploy_whitelist/blacklist in k8s are just special cases of node_selectors, specially "in" and "notin". It's all based on labels anyway.

For the replacement idea, do you mean we just set the beta selector here and just change it when we upgrade? I'm not sure I like that since there is a period of incompatibility between when we upgrade k8s and when we upgrade everything else to match.